### PR TITLE
Basic auth

### PR DESCRIPTION
**Pull request by izik1**
# Status / todo list
## database
* [x] understand basic auth
* [x] allow users to have basic auth
* [x] allow users to have verified emails
* [x] review

## routes
* [x] create a user with basic auth
* [x] log in with basic auth
* [x] verify emails
* [x] resend verification emails
* [ ] edit email after account creation
* [ ] edit password after account creation
* [ ] reset password
* [ ] change profile after account creation
* [ ] change username after account creation (unsure if wanted)
* [x] review
* [ ] documentation (kinda done)
* [ ] rate limiting auth related routes
* [ ] delete account
* [x] logout

## edge cases
* [ ] email is taken
  * [ ] oauth user already exists
  * [x] email user already exists
* [x] username is taken
* [x] username is empty
* [x] review

## concerns
* [ ] what and when is data leakage acceptable
  * [ ] saying you already have an account with an email address is fine *after* verifying it, but before...?
  * [ ] same with usernames, *where* the account came from, etc
* [x] is the design acceptable?
* [ ] lots of duplicated code, needs to be DRYed out.

# Breaking changes:
* for oauth: `POST /v1/user` -> `PUT /v1/user/me/profile`
* login token format changed, and may change *again* before a merge.

### CODE REVIEW

**Comment by izik1** [comment]
Consider replacing paseto tokens with a pair of cookies (CSRF, AUTH)

File: shared/rust/src/api/endpoints/session.rs [hunk 1]
```diff
@@ -41,3 +40,13 @@ impl ApiEndpoint for GetOAuthUrl {
     const PATH: &'static str = "/v1/session/oauth/url/{service}/{kind}";
     const METHOD: Method = Method::Get;
 }
+
+/// Delete a session (logout)
+pub struct Delete;
```
Comment (izik1) [inline shared/rust/src/api/endpoints/session.rs:45]: Not implemented

File: shared/rust/src/api/endpoints/user.rs [hunk 2]
```diff
@@ -39,3 +95,13 @@ impl ApiEndpoint for UserLookup {
     const PATH: &'static str = "/v1/user";
     const METHOD: Method = Method::Get;
 }
+
+/// Delete your account
+pub struct Delete;
```
Comment (izik1) [inline shared/rust/src/api/endpoints/user.rs:100]: Not implemented

File: shared/rust/src/api/endpoints/user.rs [hunk 3]
```diff
@@ -3,31 +3,87 @@ use super::ApiEndpoint;
...
+}
+
+/// Reset a user's password
+///
+/// # Flow (NOT IMPLEMENTED)
+/// 1. `POST` This route.
+///      * email gets sent to the included email address
+///      * recieve [`204 - No Content`](http::StatusCode::NO_CONTENT)
+/// 2. `PUT /v1/user/me/password`
+pub struct ResetPassword;
```
Comment (izik1) [inline shared/rust/src/api/endpoints/user.rs:80]: Not implemented

File: shared/rust/src/api/endpoints/user.rs [hunk 4]
```diff
@@ -3,31 +3,87 @@ use super::ApiEndpoint;
...
+/// # Register Flow
+/// 1. [`POST /v1/user`](Create)
+///     * This will send the email for verification
+/// 2. `POST` this route with the Verify request.
+///     * recieve one of:
+///         1. a 200 ([`NewSessionResponse`])
+///         2. a 401 response (`<token>` is invalid)
+///
+/// # Change email Flow (NOT CURRENTLY IMPLEMENTED)
+/// 1. `PUT /v1/user/me/email`
```
Comment (izik1) [inline shared/rust/src/api/endpoints/user.rs:48]: Not implemented

File: shared/rust/src/api/endpoints/user/profile.rs [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    type Res = UserProfile;
+    type Err = EmptyError;
+    const PATH: &'static str = "/v1/user/me/profile";
+    const METHOD: Method = Method::Get;
+}
+
+/// Create or replace your User Profile.
+///
+/// Note: Currently this can only be done at signup time,
+/// in the future `PATCH /v1/user/me/profile` will exist to handle editing your profile.
```
Comment (izik1) [inline shared/rust/src/api/endpoints/user/profile.rs:23]: Not implemented

File: backend/api/src/http/endpoints/session.rs [hunk 6]
```diff
@@ -1,186 +1,79 @@
...
 ) -> Result<HttpResponse, error::Server> {
-    // todo: make sure there isn't anything that needs to be done with the db? (eg, actually creating a session id or smth)
+    let login_ttl = settings
+        .login_token_valid_duration
+        .unwrap_or(Duration::weeks(2));
+
+    let (mask, valid_until) = match user.registration_status {
+        RegistrationStatus::New => panic!("This isn't currently possible"),
+        RegistrationStatus::Validated => (
+            SessionMask::PUT_PROFILE | SessionMask::DELETE_ACCOUNT,
```
Comment (izik1) [inline backend/api/src/http/endpoints/session.rs:36]: Do we want the `DELETE_ACCOUNT` permission here?

File: backend/api/src/http/endpoints/session/oauth.rs [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+                r#"select exists(select 1 from user_profile where user_id = $1) as "exists!""#,
+                google_auth.user_id
+            )
+            .fetch_one(&mut txn)
+            .await?;
+
+            let mask = if check_profile.exists {
+                SessionMask::GENERAL
+            } else {
+                SessionMask::PUT_PROFILE | SessionMask::DELETE_ACCOUNT
```
Comment (izik1) [inline backend/api/src/http/endpoints/session/oauth.rs:139]: Same question as in `super`

File: backend/api/src/http/endpoints/session/oauth.rs [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+
+            sqlx::query!(
+                r"insert into user_auth_google (user_id, google_id) values ($1, $2)",
+                id,
+                &claims.google_id
+            )
+            .execute(&mut txn)
+            .await?;
+
+            // todo: handle duplicate email here (return a conflict error)
```
Comment (izik1) [inline backend/api/src/http/endpoints/session/oauth.rs:162]: fix this

File: backend/api/src/http/endpoints/session/oauth.rs [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+                id,
+                &claims.email
+            )
+            .execute(&mut txn)
+            .await?;
+            (id, SessionMask::PUT_PROFILE)
+        }
+    };
+
+    let login_ttl = login_token_valid_duration.unwrap_or(Duration::weeks(2));
```
Comment (izik1) [inline backend/api/src/http/endpoints/session/oauth.rs:175]: repeated issue: sometimes the token is only is only valid for an hour, but the cookie is valid for 2 weeks at least

File: backend/api/src/http/endpoints/user.rs [hunk 10]
```diff
@@ -1,28 +1,239 @@
...
+
+    if req.password.is_empty() {
+        todo!("empty password is error");
+    }
+
+    let mut txn = db.begin().await?;
+
+    // todo: handle duplicate emails
+
+    let user = sqlx::query!(r#"insert into "user" default values returning id"#)
```
Comment (izik1) [inline backend/api/src/http/endpoints/user.rs:80]: fix this

File: backend/api/src/http/endpoints/user.rs [hunk 11]
```diff
@@ -1,28 +1,239 @@
...
+    db: Data<PgPool>,
+    req: Json<<VerifyEmail as ApiEndpoint>::Req>,
+) -> Result<HttpResponse, error::Service> {
+    let req = req.into_inner();
+
+    match req {
+        VerifyEmailRequest::Resend { email } => {
+            let mut txn = db.begin().await?;
+
+            // todo: make this more future proof and exhaustive.
```
Comment (izik1) [inline backend/api/src/http/endpoints/user.rs:146]: What does this mean?

File: backend/api/src/http/endpoints/user.rs [hunk 12]
```diff
@@ -1,28 +1,239 @@
...
+            txn.commit().await?;
+
+            Ok(HttpResponse::NoContent().into())
+        }
+
+        VerifyEmailRequest::Verify { token } => {
+            let mut txn = db.begin().await?;
+
+            // todo: make this more future proof and exhaustive.
+            // todo: handle the conflict case?
```
Comment (izik1) [inline backend/api/src/http/endpoints/user.rs:187]: Should this happen *here* or earlier? How?

File: backend/api/src/http/endpoints/user.rs [hunk 13]
```diff
@@ -1,28 +1,239 @@
...
+
+            txn.commit().await?;
+
+            Ok(HttpResponse::NoContent().into())
+        }
+
+        VerifyEmailRequest::Verify { token } => {
+            let mut txn = db.begin().await?;
+
+            // todo: make this more future proof and exhaustive.
```
Comment (izik1) [inline backend/api/src/http/endpoints/user.rs:186]: Same question as above

File: backend/api/src/http/endpoints/user.rs [hunk 14]
```diff
@@ -1,28 +1,239 @@
...
+    session.expires_at > now() and
+    (session.scope_mask & $2) = $2
+returning user_id
+"#,
+                token,
+                SessionMask::VERIFY_EMAIL.bits(),
+            )
+            .fetch_optional(&mut txn)
+            .await?
+            .ok_or_else(|| anyhow::anyhow!("this should 401"))?;
```
Comment (izik1) [inline backend/api/src/http/endpoints/user.rs:206]: Make this 401

File: backend/api/src/http/endpoints/user.rs [hunk 15]
```diff
@@ -1,28 +1,239 @@
...
+            // make sure they can't use the link, now that they're verified.
+            db::session::clear_any(&mut txn, user.user_id, SessionMask::VERIFY_EMAIL).await?;
+
+            let login_ttl = config
+                .login_token_valid_duration
+                .unwrap_or(Duration::weeks(2));
+
+            let valid_until = Utc::now() + Duration::hours(1);
+
+            let session = db::session::create(
```
Comment (izik1) [inline backend/api/src/http/endpoints/user.rs:217]: What if they already have a profile?

File: backend/api/src/http/endpoints/user.rs [hunk 16]
```diff
@@ -38,80 +249,82 @@ async fn user_lookup(
...
-            Ok(HttpResponse::Created()
-                .cookie(cookie)
-                .json(CreateSessionSuccess { csrf }))
-        }
-    }
+    let login_ttl = settings
+        .login_token_valid_duration
+        .unwrap_or(Duration::weeks(2));
+
+    let session = db::session::create(
```
Comment (izik1) [inline backend/api/src/http/endpoints/user.rs:284]: Should this route log you in?
From one perspective, you already had to verify yourself and get to this route

File: backend/api/src/db/session.rs [hunk 17]
```diff
@@ -0,0 +1,61 @@
+use chrono::{DateTime, Utc};
+use sqlx::PgConnection;
+use uuid::Uuid;
+
+use crate::token::SessionMask;
+
+#[must_use]
+fn generate_session_token() -> String {
```
Comment (mehcode) [inline backend/api/src/db/session.rs:8]: I would generate 32 _bytes_ of data and then base64 encode it. You're limiting the entropy here with just `Alphanumeric`.

File: backend/api/src/http/endpoints/user.rs [hunk 18]
```diff
@@ -1,28 +1,239 @@
...
+        .await?;
+
+    let pass = req.password;
+
+    let res = actix_web::web::block(move || {
+        let password_hasher = Argon2::default();
+
+        let salt = SaltString::generate(thread_rng());
+        password_hasher
+            .hash_password(
```
Comment (mehcode) [inline backend/api/src/http/endpoints/user.rs:91]: Use `hash_password_simple`

Comment (izik1) [reply backend/api/src/http/endpoints/user.rs:91]: I was recommended to put the parameters at a higher level (which is how I got the ones I'm using), as you may recall from a conversation on slack.

**Pull request merged**